crates/aptos-crypto/src/bls12381/bls12381_pop.rs

```diff
@@ -0,0 +1,114 @@
+use anyhow::{anyhow, Result};
+use blst::BLST_ERROR;
+use aptos_crypto_derive::{DeserializeKey, SerializeKey};
+use crate::bls12381::bls12381_keys::{BLS12381PrivateKey, BLS12381PublicKey};
+use crate::{CryptoMaterialError, Length, ValidCryptoMaterial, ValidCryptoMaterialStringExt};
+
+/// Domain separation tag (DST) for hashing a public key before computing its proof-of-possesion (PoP),
+/// which is also just a signature.
+const DST_BLS_POP_IN_G2: &[u8] = b"BLS_POP_BLS12381G2_XMD:SHA-256_SSWU_RO_POP_";
+
+#[derive(Debug, Clone, Eq, PartialEq, SerializeKey, DeserializeKey)]
+/// A proof-of-possesion (PoP) of a BLS12381 private key.
+/// This is just a BLS signature on the corresponding public key.
+pub struct BLS12381ProofOfPossession {
+    pub(crate) pop: blst::min_pk::Signature,
+}
+
+impl BLS12381ProofOfPossession {
+    /// The length of the BLS12381ProofOfPossession
+    // NOTE(Alin): We have to hardcode this here because there is no library-defined constant
+    pub const LENGTH: usize = 96;
+
+    /// Serialize a BLS12381ProofOfPossession.
+    pub fn to_bytes(&self) -> [u8; Self::LENGTH] {
+        self.pop.to_bytes()
+    }
+
+    /// Group-check the PoP: i.e., verifies the PoP is a valid group element.
+    /// WARNING: This is called implicitly in `verify`, so this function
+    /// should not be called separately for most use-cases. We leave it here just in case.
+    pub fn group_check(&self) -> Result<()> {
+        self.pop.validate(true).map_err(|e| anyhow!("{:?}", e))
+    }
+
+    /// Verifies the proof-of-possesion (PoP) of the private key corresponding to the specified
+    /// BLS12-381 public key. Implicitly, group checks the PoP and the specified public key, so
+    /// the caller is not responsible for doing it manually.
+    pub fn verify(&self, pk: &BLS12381PublicKey) -> Result<()> {
+        // CRYPTONOTE(Alin): We call the signature verification function with pk_validate set to true
+        // since we do not necessarily trust the PK we deserialized over the network whose PoP we are
+        // verifying here.
+        let result = self
+            .pop
+            .verify(true, &pk.to_bytes(), DST_BLS_POP_IN_G2, &[], &pk.pubkey, true);
+        if result == BLST_ERROR::BLST_SUCCESS {
+            Ok(())
+        } else {
+            Err(anyhow!("Proof of possession did NOT verify: {:?}", result))
+        }
+    }
+
+    /// Creates a proof-of-possesion (PoP) of the specified BLS12-381 private key. Takes the
+    /// corresponding public key as input, to avoid inefficiently recomputing it from the
+    /// private key.
+    /// WARNING: Does not group-check the PK, since this function will be typically called on
+    /// a freshly-generated key-pair or on a correctly-deserialized keypair.
+    // TODO(Alin): Are there any attack vectors worth considering here? Confirm with Zekun the under which this is called.
+    pub fn create(sk: &BLS12381PrivateKey, pk: &BLS12381PublicKey) -> BLS12381ProofOfPossession {
```

Comment:
do we need to pass both private and public key here? it seems more intuitive to pass private key and derive the public key inside

Reply:
We could, but computing the PK will slow down the computation of the PoP. Plus, usually, one already has the PK before they call this function if they use the `KeyPair` pattern. Ideally, the PoP would be included in the `KeyPair` and computed when the key is generated? Its verification would still have to be done explicitly by whomever relies on the public key for verifying (multi)sigs

Reply:
On the other hand, recomputing the PK in the function would not be the end of the world, since this is done only once when a validator generates its key-pair? 

But I don't know enough about whether these key-pairs expire and new ones are re-generated often. In that case, we might want to avoid recomputation.

I can certainly provide both functionalities via separate functions.

Reply:
we should only expect creating it once, so recompute it shouldn't be too bad

Reply:
So I'll change `BLS12381ProofOfPossesion::create` to just take the `sk` and add a `BLS12381ProofOfPossesion::create_with_pubkey` that additionally takes the `pk`